### DISCUSSION TITLE
association_rules over PyEclat output

### DISCUSSION THREAD

**Discussion by catarinarurbano** [rasbt/mlxtend#959] [Q&A] [answered]
Hi Sebastian,

Hope you are well.

Can the **association_rules** be used over the output of [pyEclat](https://pypi.org/project/pyECLAT/) package, which is the frequent itemsets, to generate the association rules?

**Comment by rasbt** [comment, marked as answer]
In mlxtend, there are two steps for association rules: the frequent itemset generation (currently via apriori, fp-growth, or fp-max), and then the `association_rules` function you mentioned. Since the `association_rules` is tailored to the output format of  apriori, fp-growth, and fp-max, the pyEclat outputs may not be directly compatible.

However, looking at the pyEclat package (which was released about 2 years ago) it seems that the author made it compatible with association_rules (implemented like 5 years ago in mlxtend). I.e., the `eclat_class.df_bin` output here https://github.com/jeffrichardchemistry/pyECLAT/blob/master/examples/example.ipynb looks compatible but I haven't tried.

**Reply by catarinarurbano** [reply]
Thanks for your quick response, I'm trying to use the association_rules function as the pyEclat library only outputs the frequent itemsets, thus missing the rule generation part. However, the final output for frequent itemsets is **get_ECLAT_supports_all**. So I think it's not in the same format, unfortunately. So, I followed this [Implementation](https://hands-on.cloud/implementation-of-eclat-algorithm-using-python/) to get a dataframe as close as possible to the needed output format for the **association_rules** function. Nevertheless, I think it's not the same as the frequent itemset generation since the itemsets aren't in the form (item1, item2,...,itemN). Do you know any way to make it compatible?

In addition, as the Eclat algorithm only uses the support metric, the parameter support_only should be set to True, right?

Update: I found in a post [TowardsDataScience](https://towardsdatascience.com/the-eclat-algorithm-8ae3276d2d17) that they consider **get_ECLAT_supports_all** as the rules itself, which doesn't make much sense...

Many thanks for your support on the matter.

**Reply by rasbt** [reply]
Oh I see now. The output looks slightly different because they use "&" instead of a comma to separate the items. Btw. the `support_only=True` is a bit misleading. Here, it means that the output only contains the `support` and other values will not be calculated. It is expected that the input dataframe only contains frequent itemsets and the support.

I just gave it a quick try and the following should work:

```python
from pyECLAT import Example1, Example2
from pyECLAT import ECLAT
import pandas as pd

ex1 = Example1().get()
eclat_instance = ECLAT(data=ex1, verbose=True)

get_ECLAT_indexes, get_ECLAT_supports = eclat_instance.fit(min_support=0.08,
                                                           min_combination=1,
                                                           max_combination=3,
                                                           separator=' & ',
                                                           verbose=True)

df = pd.DataFrame(get_ECLAT_supports.items(),columns=['itemsets', 'support'])
df
```

![Screen Shot 2022-07-21 at 8 53 53 PM](https://user-images.githubusercontent.com/5618407/180345225-b4d15511-8601-48ec-a415-397e9447e8e8.png)

```python
new_column = []

for row in df['itemsets']:
    r = row.split('&')
    r = tuple(i.strip() for i in r)
    new_column.append(r)

df['itemsets'] = pd.Series(new_column)

df
```
![Screen Shot 2022-07-21 at 8 54 27 PM](https://user-images.githubusercontent.com/5618407/180345288-5e07efe7-4a8f-4418-9a01-2ebcce2f6792.png)


```python
from mlxtend.frequent_patterns import association_rules

association_rules(df, metric="confidence", min_threshold=0.1)
```

![Screen Shot 2022-07-21 at 8 54 52 PM](https://user-images.githubusercontent.com/5618407/180345344-37bc4658-de92-46d1-9981-a07fdd8f1ae4.png)



Please let me know if it works (and if it does, it would be nice to check the answer as correct -- it might be useful as a future reference for others.)

PS: I also agree with you that the TowardsDataScience article is a bit weird 😆

**Reply by catarinarurbano** [reply]
Sebastian,

thanks a lot for your help. It solved my issue and clarified my doubts. 
My question regarding the **support_only=True** was just because the ECLAT algorithm only uses the support metric (it lacks other interestingness measures such as lift and confidence), so I think the most correct thing would be not to present the values of these measures in the output. 

Thanks again for your support.
Have a great day!

**Reply by rasbt** [reply]
Yay! Glad to hear that it worked!
